From 000197c79912936bc2797795fc9eb3aa30aa4349 Mon Sep 17 00:00:00 2001
From: Cheng Jian <cj.chengjian@huawei.com>
Date: Sat, 29 May 2021 14:43:10 +0800
Subject: [PATCH] livepatch/ppc64: Sample testcase fix ppc64
Patch-mainline: Not yet, from openEuler
References: bsn#22
openEuler-commit: 000197c79912936bc2797795fc9eb3aa30aa4349


hulk inclusion
category: feature
bugzilla: 51924
CVE: NA

---------------------------

The previous sample use case did not consider the APC and function
descriptors of PPC64

Signed-off-by: Cheng Jian <cj.chengjian@huawei.com>
Reviewed-By: Xie XiuQi <xiexiuqi@huawei.com>
Signed-off-by: yangerkun <yangerkun@huawei.com>

Signed-off-by: Dong Kai <dongkai11@huawei.com>

Signed-off-by: Ye Weihua <yeweihua4@huawei.com>
Reviewed-by: Yang Jihong <yangjihong1@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 samples/livepatch/livepatch-sample.c | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/samples/livepatch/livepatch-sample.c b/samples/livepatch/livepatch-sample.c
index cd76d7ebe598..2ffeae67dc14 100644
--- a/samples/livepatch/livepatch-sample.c
+++ b/samples/livepatch/livepatch-sample.c
@@ -10,6 +10,9 @@
 #include <linux/module.h>
 #include <linux/kernel.h>
 #include <linux/livepatch.h>
+#ifdef CONFIG_PPC64
+#include <asm/code-patching.h>
+#endif
 
 /*
  * This (dumb) live patch overrides the function that prints the
@@ -38,7 +41,11 @@ static int livepatch_cmdline_proc_show(struct seq_file *m, void *v)
 
 static struct klp_func funcs[] = {
 	{
+#ifdef CONFIG_PPC64
+		.old_name = ".cmdline_proc_show",
+#else
 		.old_name = "cmdline_proc_show",
+#endif
 		.new_func = livepatch_cmdline_proc_show,
 	}, { }
 };
@@ -57,6 +64,10 @@ static struct klp_patch patch = {
 
 static int livepatch_init(void)
 {
+#ifdef CONFIG_PPC64
+	patch.objs[0].funcs[0].new_func =
+		(void *)ppc_function_entry((void *)livepatch_cmdline_proc_show);
+#endif
 	return klp_enable_patch(&patch);
 }
 
-- 
2.26.2

